### ISSUE TITLE
How to train VA-VAE?

### ISSUE THREAD

**Issue by lavinal712** [hustvl/LightningDiT#1]
Can you release the code? Thanks for your works!

**Comment by JingfengYao** [comment]
The training code for VA-VAE is primarily based on the autoencoder training code from [LDM](https://github.com/CompVis/latent-diffusion/tree/main). Implementing it should be relatively straightforward, as described in Section 3 of the paper. We are currently considering the most concise way to release it, such as forking or something else.

**Comment by lavinal712** [comment]
Thanks! I tried to reproduce your code, but I found that the vf_loss does not converge easily. After training for 1000 steps, the model collapsed, and the output turned into solid-color images. Therefore, I would like to see more details in the code.

**Comment by gkakogeorgiou** [comment]
Hi! Thanks for the great work! I noticed similar behavior to what @lavinal712 observed. Releasing the VA-VAE training code could help us better understand the process.

**Comment by lavinal712** [comment]
"""
if z_prime is not None:
                if self.vf_loss_type == "combined_v3":
                    mcos_loss = nn.functional.cosine_similarity(z_prime, features, dim=-1)
                    mcos_loss = nn.functional.relu(1 - self.cos_margin - mcos_loss)
                    mcos_loss = torch.mean(mcos_loss)
                    
                    z_prime = nn.functional.normalize(z_prime, dim=-1)
                    features = nn.functional.normalize(features, dim=-1)
                    
                    cossim_z_prime = torch.matmul(z_prime, z_prime.transpose(1, 2))
                    cossim_features = torch.matmul(features, features.transpose(1, 2))
                    
                    mdms_loss = torch.abs(cossim_z_prime - cossim_features)
                    mdms_loss = nn.functional.relu(mdms_loss - self.distmat_margin)
                    mdms_loss = torch.mean(mdms_loss)
                    
                    vf_loss = mcos_loss + mdms_loss
                else:
                    raise ValueError(f"Unknown vf_loss_type: {self.vf_loss_type}")
                
                if self.adaptive_vf:
                    try:
                        vf_weight = self.calculate_adaptive_weight(nll_loss, vf_loss, last_layer=encoder_last_layer)
                    except RuntimeError:
                        assert not self.training
                        vf_weight = torch.tensor(1.0)
                    vf_loss = vf_weight * vf_loss
            else:
                vf_loss = torch.tensor(0.0)
"""

@gkakogeorgiou @JingfengYao Is there any problem of the vf_loss?

**Comment by JingfengYao** [comment]
I'm not sure about the shapes and features of your z_prime. Judging from your ``mcos_loss = nn.functional.cosine_similarity(z_prime, features, dim=-1)`` and ``z_prime = nn.functional.normalize(z_prime, dim=-1)``, are you placing the channel dimension in the last dimension?

Here are my implementations:
```
elif self.vf_loss_type == "combined_v3":
    # we also give a 0.25 margin to image dist mat loss
    z_flat = rearrange(z, 'b c h w -> b c (h w)')
    aux_feature_flat = rearrange(aux_feature, 'b c h w -> b c (h w)')
    z_norm = torch.nn.functional.normalize(z_flat, dim=1)
    aux_feature_norm = torch.nn.functional.normalize(aux_feature_flat, dim=1)
    z_cos_sim = torch.einsum('bci,bcj->bij', z_norm, z_norm)
    aux_feature_cos_sim = torch.einsum('bci,bcj->bij', aux_feature_norm, aux_feature_norm)
    diff = torch.abs(z_cos_sim - aux_feature_cos_sim)
    vf_loss_1 = torch.nn.functional.relu(diff-self.distmat_margin).mean()

    # margin_cos
    vf_loss_2 = torch.nn.functional.relu(1 - self.cos_margin - torch.nn.functional.cosine_similarity(aux_feature, z)).mean()
    vf_loss = vf_loss_1 + vf_loss_2
```

By the way, here are 2 small issues:
1. We have not imposed an upper limit on the adaptive weighting for the vf loss, hence the upper bound value in ``d_weight = torch.clamp(d_weight, 0.0, 1e4).detach()`` is set to a relatively large number with ``d_weight = torch.clamp(d_weight, 0.0, 1e8).detach()`` specifically for vf.
2. https://github.com/CompVis/latent-diffusion/blob/main/ldm/models/autoencoder.py#L386 There is a particular aspect in the LDM code that requires attention. Please ensure that the newly added projection layer for Z is correctly incorporated into the optimizer.

**Comment by lavinal712** [comment]
@JingfengYao Thanks for your clarification.
The shape of z_prime is correct. I intentionally placed the channels dimension in the last dimension to align with the dimensionality of DINOv2. The code snippet is as follows:

```
z = posterior.sample()
z_prime = z.flatten(2).transpose(1, 2)
z_prime = self.proj(z_prime)
```

This ensures compatibility with DINOv2's expected input format.

And thank you for pointing this out! I agree that the connection layer should be included in the optimizer. I will update the code to ensure that the connection layer's parameters are properly added to the optimizer for training.

**Comment by lavinal712** [comment]
Well, the model collapses again and I do not know why. Here is the code: https://github.com/[lavinal712/VA-VAE](https://github.com/lavinal712/VA-VAE)

**Comment by JingfengYao** [comment]
Could you please provide your tensorboard logs?

**Comment by lavinal712** [comment]
![image](https://github.com/user-attachments/assets/0d30d08c-3504-4a93-8843-2fb0395d28e9)
![image](https://github.com/user-attachments/assets/58d011a3-9874-4836-814e-685b462fa296)
The kl loss is extremely high.

**Comment by JingfengYao** [comment]
May I ask how many GPUs you use for training?

**Comment by JingfengYao** [comment]
Seems I found the possible reason. Here are my reproductions:

1. I clone your codes and run it with one GPU (I am not sure your starting command. In VA-VAE implementations, I modify to use ``torchrun`` for multi-gpu training)

    I start with 
    ```
    python main.py --base configs/autoencoder/vavae_f16d32.yaml --train --gpu 0,
    ```
    and get a similar problem with you. My KL loss is also extremely high. That is because of this:
    ```
    if opt.scale_lr:
        model.learning_rate = accumulate_grad_batches * ngpu * bs * base_lr
        print(
            "Setting learning rate to {:.2e} = {} (accumulate_grad_batches) * {} (num_gpus) * {} (batchsize) * {:.2e} (base_lr)".format(
                model.learning_rate, accumulate_grad_batches, ngpu, bs, base_lr))
    ```
    in my setting, lr is set to ``2.3e-3``, which means a much larger learning rate and smaller batch size than our real training. We use fixed ``1e-4`` lr and 256 or 512 batch size for training.

2. Then I modify the command as following and log vf_loss before adaptive weighting:

    ```
    python main.py --base configs/autoencoder/vavae_f16d32.yaml --train --gpu 0, --scale_lr false
    ```
    The training becomes more stable:
    <img width="1546" alt="1736335677520" src="https://github.com/user-attachments/assets/a4154861-0ea5-4d1b-8d98-7bce4fd06aa0" />

Hope this helps.

**Comment by lavinal712** [comment]
Thanks! Is it normal for the KL loss to gradually increase?

**Comment by lavinal712** [comment]
> May I ask how many GPUs you use for training and your starting command? 请问您训练时使用了多少 GPU 以及您的启动命令是什么？@lavinal712

4 GPUs, 

```
python main.py --b configs/autoencoder/vavae_f16d32.yaml --t --gpu 0,1,2,3
```

**Comment by lavinal712** [comment]
The batch size I used for training the VAE is too small. How many GPUs did you use for training in your paper? Are there any methods to increase the batch size?

**Comment by JingfengYao** [comment]
Yes, the weight of the KL loss is relatively small among various losses. A tremendous KL loss might impact the generation performance. We utilized 32/64 GPUs to train the VA-VAE. Perhaps you could experiment with mixed precision training, checkpointing, and gradient accumulation (which seems to have been already employed) to increase the batch size.

**Comment by lavinal712** [comment]
> The training code for VA-VAE is primarily based on the autoencoder training code from [LDM](https://github.com/CompVis/latent-diffusion/tree/main). Implementing it should be relatively straightforward, as described in Section 3 of the paper. We are currently considering the most concise way to release it, such as forking or something else.VA-VAE 的训练代码主要基于 LDM 的自编码器训练代码。实现起来应该相对简单，如论文第 3 节所述。我们目前正在考虑以最简洁的方式发布它，例如分叉或其他方式。

Are you still considering releasing the code for va-vae?

**Comment by JingfengYao** [comment]
@lavinal712 @gkakogeorgiou 
Sorry to reply late. We have released our training codes, reference logs and all experimental checkpoints. Please check them [here](https://github.com/hustvl/LightningDiT/tree/main/vavae).

**Comment by lavinal712** [comment]
I have some questions about normalization: During training, we normalize the images with a mean of 0.5 and a variance of 0.5, but in DINO, the input images are typically normalized using the mean and variance of ImageNet. In your code, I didn't notice any place where re-normalization is performed. Will this have an impact on the model?

**Comment by JingfengYao** [comment]
That's a great detail—thank you for pointing it out. Based on our experimental experience, this normalization detail may have a relatively minor impact on performance.
